### PR TITLE
Wire sharing UI up to model user API.

### PR DESCRIPTION
**Pull request by kadams54**

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6714/

**Comment by kadams54** [comment]
## QA

1. `make clean-gui fast-babel gui run`
2. `guiproxy -config 'shareFlag: true' -controller jimm.jujucharms.com:443`
3. Head to http://localhost:8042.
4. On the canvas, you should see the sharing button in the header. Click on it. You should see a list of all users with access to the model.

A fun one to do this on is `jujugui-org`. Unfortunately the UI is not as pretty as it could be, particularly when the API doesn't return certain data (e.g., a displayName for all users), but it's a start.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6715/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6716/

File: jujugui/static/gui/src/app/components/sharing/sharing.js [hunk 1]
```diff
@@ -89,7 +104,13 @@ YUI.add('sharing', function() {
       @returns {Array} An array of markup objects for each user.
     */
     _generateUsersWithAccess: function() {
-      // TODO kadams54: display a spinner instead when loading data.
+      if (this.state.loadingUsers) {
+        return (
+          <div className="sharing__loading">
+            <juju.components.Spinner />
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/sharing/sharing.js]: This component needs to be added in the `requires`

File: jujugui/static/gui/src/app/store/env/api.js [hunk 2]
```diff
@@ -2861,6 +2861,59 @@ YUI.add('juju-env-api', function(Y) {
...
+          callback(data.error, []);
+          return;
+        }
+
+        const results = data.response.results;
+        const users = results.map((result, index) => {
+          const err = result.error && result.error.message;
+          result = result.result || {};
+          return {
+            name: result.user,
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/api.js]: Is this a user tag? If so you'll probably want to `tags.parse()` like with other methods here to maintain consistency.

File: jujugui/static/gui/src/app/store/env/controller-api.js [hunk 3]
```diff
@@ -570,6 +570,7 @@ YUI.add('juju-controller-api', function(Y) {
       });
     },
 
+
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: extra line?

**Comment by hatched** [comment]
QA Notes:

The sharing button is visible in the profile view, clicking it throws an error because we aren't connected to a model.

File: jujugui/static/gui/src/app/components/sharing/sharing.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/sharing/sharing.js]: Done at https://github.com/juju/juju-gui/pull/2411

File: jujugui/static/gui/src/app/store/env/api.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/api.js]: It's not a user tag. Anyway, this handling has been slightly changed at https://github.com/juju/juju-gui/pull/2411

File: jujugui/static/gui/src/app/store/env/controller-api.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/controller-api.js]: Done at https://github.com/juju/juju-gui/pull/2411

**Comment by frankban** [comment]
This is superseded by https://github.com/juju/juju-gui/pull/2411 in an attempt to include the functionality in 2.3.0. Please refer to that merge proposal, which also includes hiding model actions in the profile view.

**Pull request closed without merging**